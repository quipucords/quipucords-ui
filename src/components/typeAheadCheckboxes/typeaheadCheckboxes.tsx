/**
 * Implements a multi-select typeahead component with checkboxes, enabling users to filter and select multiple options.
 * Utilizes PatternFly for UI consistency and accessibility.
 * @module typeaheadCheckboxes
 */
import React from 'react';
import {
  Select,
  SelectOption,
  SelectList,
  SelectOptionProps,
  MenuToggle,
  MenuToggleElement,
  TextInputGroup,
  TextInputGroupMain,
  TextInputGroupUtilities,
  Button
} from '@patternfly/react-core';
import { TimesIcon } from '@patternfly/react-icons/';

export interface ITypeaheadCheckboxesProps {
  onChange: (selections: string[]) => void;
  options: { value: string; label: string }[];
  selectedOptions?: string[];
  placeholder?: string;
}

export const TypeaheadCheckboxes: React.FC<ITypeaheadCheckboxesProps> = ({
  onChange,
  options,
  selectedOptions = [],
  placeholder = '0 items selected'
}) => {
  const [isOpen, setIsOpen] = React.useState(false);
  const [inputValue, setInputValue] = React.useState<string>('');
  const [selected, setSelected] = React.useState<string[]>(selectedOptions || []);
  const [selectOptions, setSelectOptions] = React.useState<SelectOptionProps[]>(options);
  const [focusedItemIndex, setFocusedItemIndex] = React.useState<number | null>(null);
  const [activeItem, setActiveItem] = React.useState<string | null>(null);
  const [activePlaceholder, setActivePlaceholder] = React.useState(placeholder);
  const textInputRef = React.useRef<HTMLInputElement>();

  React.useEffect(() => {
    setSelectOptions(options);
  }, [options]);

  React.useEffect(() => {
    let newSelectOptions: SelectOptionProps[] = options;

    // Filter menu items based on the text input value when one exists
    if (inputValue) {
      newSelectOptions = options.filter(menuItem =>
        String(menuItem.label).toLowerCase().includes(inputValue.toLowerCase())
      );

      // When no options are found after filtering, display 'No results found'
      if (!newSelectOptions.length) {
        newSelectOptions = [
          {
            isDisabled: false,
            children: `No results found for "${inputValue}"`,
            value: 'no results'
          }
        ];
      }

      // Open the menu when the input value changes and the new value is not empty
      if (!isOpen) {
        setIsOpen(true);
      }
    }

    setSelectOptions(newSelectOptions);
    setFocusedItemIndex(null);
    setActiveItem(null);
  }, [inputValue, isOpen, options]);

  const handleMenuArrowKeys = (key: string) => {
    let indexToFocus;

    if (isOpen) {
      if (key === 'ArrowUp') {
        // When no index is set or at the first index, focus to the last, otherwise decrement focus index
        if (focusedItemIndex === null || focusedItemIndex === 0) {
          indexToFocus = selectOptions.length - 1;
        } else {
          indexToFocus = focusedItemIndex - 1;
        }
      }

      if (key === 'ArrowDown') {
        // When no index is set or at the last index, focus to the first, otherwise increment focus index
        if (focusedItemIndex === null || focusedItemIndex === selectOptions.length - 1) {
          indexToFocus = 0;
        } else {
          indexToFocus = focusedItemIndex + 1;
        }
      }

      setFocusedItemIndex(indexToFocus);
      const focusedItem = selectOptions.filter(option => !option.isDisabled)[indexToFocus];
      setActiveItem(
        `select-multi-typeahead-checkbox-${String(focusedItem.value).replace(' ', '-')}`
      );
    }
  };

  const onInputKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    const enabledMenuItems = selectOptions.filter(menuItem => !menuItem.isDisabled);
    const [firstMenuItem] = enabledMenuItems;
    const focusedItem = focusedItemIndex ? enabledMenuItems[focusedItemIndex] : firstMenuItem;

    switch (event.key) {
      // Select the first available option
      case 'Enter':
        if (!isOpen) {
          setIsOpen(prevIsOpen => !prevIsOpen);
        } else if (isOpen && focusedItem.value !== 'no results') {
          onSelect(focusedItem.value as string);
        }
        break;
      case 'Tab':
      case 'Escape':
        setIsOpen(false);
        setActiveItem(null);
        break;
      case 'ArrowUp':
      case 'ArrowDown':
        event.preventDefault();
        handleMenuArrowKeys(event.key);
        break;
    }
  };

  const onToggleClick = () => {
    setIsOpen(!isOpen);
  };

  const onTextInputChange = (_event: React.FormEvent<HTMLInputElement>, value: string) => {
    setInputValue(value);
  };

  const onSelect = (value: string) => {
    if (value && value !== 'no results') {
      const newSelected = selected.includes(value)
        ? selected.filter(selection => selection !== value)
        : [...selected, value];
      onChange(newSelected);
      setSelected(newSelected);
    }

    textInputRef.current?.focus();
  };

  React.useEffect(() => {
    setActivePlaceholder(selected.length ? `${selected.length} items selected` : placeholder);
  }, [selected, placeholder]);

  const toggle = (toggleRef: React.Ref<MenuToggleElement>) => (
    <MenuToggle
      variant="typeahead"
      onClick={onToggleClick}
      innerRef={toggleRef}
      isExpanded={isOpen}
      isFullWidth
    >
      <TextInputGroup isPlain>
        <TextInputGroupMain
          value={inputValue}
          onClick={onToggleClick}
          onChange={onTextInputChange}
          onKeyDown={onInputKeyDown}
          id="multi-typeahead-select-checkbox-input"
          autoComplete="off"
          innerRef={textInputRef}
          placeholder={activePlaceholder}
          {...(activeItem && { 'aria-activedescendant': activeItem })}
          role="combobox"
          isExpanded={isOpen}
          aria-controls="select-multi-typeahead-checkbox-listbox"
        />
        <TextInputGroupUtilities>
          {selected.length > 0 && (
            <Button
              variant="plain"
              onClick={() => {
                setInputValue('');
                setSelected([]);
                textInputRef?.current?.focus();
              }}
              aria-label="Clear input value"
            >
              <TimesIcon aria-hidden />
            </Button>
          )}
        </TextInputGroupUtilities>
      </TextInputGroup>
    </MenuToggle>
  );

  return (
    <Select
      role="menu"
      id="multi-typeahead-checkbox-select"
      isOpen={isOpen}
      selected={selected}
      onSelect={(ev, selection) => onSelect(selection as string)}
      onOpenChange={() => setIsOpen(false)}
      toggle={toggle}
    >
      <SelectList id="select-multi-typeahead-checkbox-listbox">
        {selectOptions.map((option, index) => (
          <SelectOption
            {...(!option.isDisabled && { hasCheckbox: true })}
            isSelected={selected.includes(option.value)}
            key={option.value || option.children}
            isFocused={focusedItemIndex === index}
            className={option.className}
            id={`select-multi-typeahead-${String(option.value).replace(' ', '-')}`}
            // Pass children before spreading ...option so it can be overridden by the consumer
            // eslint-disable-next-line react/no-children-prop
            children={option.label}
            {...option}
            ref={null}
          />
        ))}
      </SelectList>
    </Select>
  );
};
