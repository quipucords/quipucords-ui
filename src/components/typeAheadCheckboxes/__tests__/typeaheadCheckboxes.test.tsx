import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { shallowComponent } from '../../../../config/jest.setupTests';
import { TypeaheadCheckboxes } from '../typeaheadCheckboxes';

describe('TypeaheadCheckboxes', () => {
  it('should render a basic component', async () => {
    const props = {
      options: [{ value: 'ipsum', label: 'Lorem' }]
    };
    const component = await shallowComponent(<TypeaheadCheckboxes {...props} />);
    expect(component).toMatchSnapshot('basic');
  });

  it('should clear selections and call onChange with [] when clear button is clicked', async () => {
    const mockOnChange = jest.fn();
    render(
      <TypeaheadCheckboxes
        options={[
          { value: 'alpha', label: 'Alpha' },
          { value: 'beta', label: 'Beta' }
        ]}
        selectedOptions={['alpha', 'beta']}
        onChange={mockOnChange}
      />
    );

    const clearButton = await screen.findByRole('button', { name: /clear input value/i });
    await userEvent.click(clearButton);

    expect(mockOnChange).toHaveBeenCalledWith([]);
  });

  it('should filter options, select one, and close dropdown when maxSelections is 1', async () => {
    const onChange = jest.fn();

    render(
      <TypeaheadCheckboxes
        options={[
          { value: 'alpha', label: 'Alpha' },
          { value: 'beta', label: 'Beta' }
        ]}
        selectedOptions={[]}
        onChange={onChange}
        maxSelections={1}
      />
    );

    const input = screen.getByRole('combobox');
    await userEvent.type(input, 'alp');

    const alphaOption = await screen.findByText('Alpha');
    await userEvent.click(alphaOption);

    expect(onChange).toHaveBeenCalledWith(['alpha']);

    // Wait for the menu to disappear
    await waitFor(() => expect(screen.queryByRole('menu')).not.toBeInTheDocument());
  });

  it('should filter options, select multiple, and call onChange twice (controlled)', async () => {
    const user = userEvent.setup();
    const options = [
      { value: 'alpha', label: 'Alpha' },
      { value: 'beta', label: 'Beta' },
      { value: 'gamma', label: 'Gamma' }
    ];

    /**
     *
     */
    const Wrapper = () => {
      const [selected, setSelected] = React.useState<string[]>([]);
      return (
        <TypeaheadCheckboxes options={options} selectedOptions={selected} onChange={setSelected} maxSelections={2} />
      );
    };

    render(<Wrapper />);

    const input = screen.getByRole('combobox');
    await user.type(input, 'a');

    const alphaOption = await screen.findByText('Alpha');
    await user.click(alphaOption);

    // Re-open the dropdown before selecting 'Beta'
    await user.click(input);
    await user.clear(input);
    await user.type(input, 'bet');
    const betaOption = await screen.findByText('Beta');
    await user.click(betaOption);

    // Now both should be selected
    expect(screen.getByPlaceholderText('2 items selected')).toBeInTheDocument();

    // Ensure the dropdown closes when maxSelections is reached
    await waitFor(() => expect(screen.queryByRole('menu')).not.toBeInTheDocument());
  });

  it('shows "No results found" when filter yields no matches', async () => {
    render(
      <TypeaheadCheckboxes
        options={[
          { value: 'alpha', label: 'Alpha' },
          { value: 'beta', label: 'Beta' }
        ]}
        selectedOptions={[]}
      />
    );

    const input = screen.getByRole('combobox');
    await userEvent.type(input, 'zzz');

    expect(screen.getByText(/No results found/i)).toBeInTheDocument();
  });
});
