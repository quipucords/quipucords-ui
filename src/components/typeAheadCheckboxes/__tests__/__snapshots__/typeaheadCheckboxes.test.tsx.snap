// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeaheadCheckboxes should render a basic component: basic 1`] = `
<Select
  id="multi-typeahead-checkbox-select-with-search"
  isOpen={false}
  onOpenChange={[Function]}
  onSelect={[Function]}
  role="menu"
  selected={[]}
  toggle={[Function]}
>
  <SelectList
    id="select-multi-typeahead-checkbox-listbox-with-search"
  />
</Select>
`;
